Language/Java/Transformer.hs

```diff
@@ -0,0 +1,391 @@
+{-# LANGUAGE LambdaCase #-}
+
+module Language.Java.Transformer (analyze, IdentCollection (..)) where
+
+import Data.Bifunctor (Bifunctor (second))
+import Data.Maybe (mapMaybe)
+import Language.Java.Syntax
+
+class Transform a where
+  analyze :: IdentCollection -> a Parsed -> a Analyzed
+
+-- | data type used to collect the identifiers that are currently in scope
+data IdentCollection = IdentCollection
+  { fields :: [Ident],
+    formalParams :: [Ident],
+    localVars :: [Ident]
+  }
+
+addFields :: IdentCollection -> ClassDecl Parsed -> IdentCollection
+addFields identCollection classDecl = identCollection {fields = fieldsFromClassDecl classDecl ++ fields identCollection}
+
+addFormalParams :: IdentCollection -> [Ident] -> IdentCollection
+addFormalParams identCollection idents = identCollection {formalParams = idents ++ formalParams identCollection}
+
+addLocalVars :: IdentCollection -> [Ident] -> IdentCollection
+addLocalVars identCollection idents = identCollection {localVars = idents ++ localVars identCollection}
+
+instance Transform ClassDecl where
+  analyze scope classDecl@(ClassDecl src modifiers ident typeParams mbRefType refTypes classBody) =
+    let newScope = addFields scope classDecl
+     in ClassDecl
+          src
+          (map (analyze newScope) modifiers)
+          ident
+          typeParams
+          mbRefType
+          refTypes
+          (analyze newScope classBody)
+  analyze scope recordDecl@(RecordDecl src modifiers ident typeParams recordFieldDecls refTypes classBody) =
+    let newScope = addFields scope recordDecl
+     in RecordDecl
+          src
+          (map (analyze newScope) modifiers)
+          ident
+          typeParams
+          recordFieldDecls
+          refTypes
+          (analyze newScope classBody)
+  analyze scope enumDecl@(EnumDecl src modifiers ident refTypes enumBody) =
+    let newScope = addFields scope enumDecl
+     in EnumDecl
+          src
+          (map (analyze newScope) modifiers)
+          ident
+          refTypes
+          (analyze newScope enumBody)
+
+instance Transform MemberDecl where
+  analyze scope (FieldDecl span modifiers type_ varDecls) = FieldDecl span (map (analyze scope) modifiers) type_ (map (analyze scope) varDecls)
+  analyze scope methodDecl@(MethodDecl span modifiers typeParams mbType ident formalParams exceptionTypes mbExp methodBody) =
+    MethodDecl
+      span
+      (map (analyze scope) modifiers)
+      typeParams
+      mbType
+      ident
+      (map (analyze scope) formalParams)
+      exceptionTypes
+      (fmap (analyze scope) mbExp)
```

Comment:
Wofür wird die `mbExp` genutzt?

Reply:
Nicht sicher. In der [Javaspezifikation](https://docs.oracle.com/javase/specs/jls/se7/html/jls-8.html#jls-8.4) finde ich diesen Teil nicht . Im Pretty hat es den Bezeichner def und wird mit der Methode ppDefault weiterverarbeitet. Also vielleicht etwas in die Richtung [defaultmethods](https://docs.oracle.com/javase/tutorial/java/IandI/defaultmethods.html). Habe das aber getestet und der Parser wirft dann den Fehler "unexpected KW_Default".

Reply:
das ist um einen default Value bei der Deklaration von Annotations zu setzen. [UsingDefaultValues](http://www.java2s.com/Tutorial/Java/0020__Language/UsingDefaultValues.htm). Der Parser kann auch normale Interfaces damit parsen, obwohl es ein Compilerfehler wäre.

Reply:
Im Parser muss nach dem Keyword default eine assignExp folgen.
`assignExp = try switchExp <|> try methodRef <|> try lambdaExp <|> try assignment <|> condExp` 
Laut [Javaspecification ](https://docs.oracle.com/javase/specs/jls/se20/html/jls-9.html#jls-9.6.2) wird default gefolgt von einem ElementValue. 
DefaultValue:
default [ElementValue](https://docs.oracle.com/javase/specs/jls/se20/html/jls-9.html#jls-ElementValue)
ElementValue:
[ConditionalExpression](https://docs.oracle.com/javase/specs/jls/se20/html/jls-15.html#jls-ConditionalExpression)
[ElementValueArrayInitializer](https://docs.oracle.com/javase/specs/jls/se20/html/jls-9.html#jls-ElementValueArrayInitializer)
[Annotation](https://docs.oracle.com/javase/specs/jls/se20/html/jls-9.html#jls-Annotation)

Und in IntelliJ gibt es mir den Hinweis "Attribute value must be constant"

Reply:
Erstell bitte mal einen Issue mit diesen Informationen. Ggf. könnte man später mal den AST umbauen, um hier mehr Strukturinformationen zu haben. Das heißt, hier kann eigentlich gar keine allgemeine Expression stehen.